README.md
# crypto_rsassa_pss

[![Build Status](https://travis-ci.org/potatosalad/erlang-crypto_rsassa_pss.svg?branch=master)](https://travis-ci.org/potatosalad/erlang-crypto_rsassa_pss) [![Hex.pm](https://img.shields.io/hexpm/v/crypto_rsassa_pss.svg)](https://hex.pm/packages/crypto_rsassa_pss)

RSASSA-PSS Public Key Cryptographic Signature Algorithm for Erlang and Elixir.

Implemented closely after the [crypto-pubkey](https://github.com/vincenthz/hs-crypto-pubkey) project for Haskell.

See [RFC #3447](https://tools.ietf.org/html/rfc3447) and [RFC #4506](https://tools.ietf.org/html/rfc4056).

## Installation

Add `crypto_rsassa_pss` to your project's dependencies in `mix.exs`

```elixir
defp deps do
  [
    {:crypto_rsassa_pss, "~> 2.0.0"}
  ]
end
```

Add `crypto_rsassa_pss` to your project's dependencies in your `Makefile` for [`erlang.mk`](https://github.com/ninenines/erlang.mk) or the following to your `rebar.config`

```erlang
{deps, [
  {crypto_rsassa_pss, ".*", {git, "git://github.com/potatosalad/erlang-crypto_rsassa_pss.git", {branch, "master"}}}
]}.
```

## Usage

First, let's generate a RSA key using `openssl`.

```bash
$ openssl genrsa -out rsa.pem 2048
Generating RSA private key, 2048 bit long modulus
...................+++
...............................................+++
e is 65537 (0x10001)
```

Now let's use our key to sign and verify a message.

### Erlang

```erl
1> {ok, Data} = file:read_file("rsa.pem").
{ok,<<"-----BEGIN RSA PRIVATE KEY-----\nMIIEowIBAAKCAQEAsBdtC3nJpzZ9tbcEIEBY4DPUdkiCLwjbHyZFzL+G72ambWsy\nkgnlSdW3JKR"...>>}
2> [PEMEntry | _] = public_key:pem_decode(Data).                
[{'RSAPrivateKey',<<48,130,4,163,2,1,0,2,130,1,1,0,176,23,
                    109,11,121,201,167,54,125,181,183,4,
                    32,64,...>>,
                  not_encrypted}]
3> #'RSAPrivateKey'{modulus=Modulus, publicExponent=PublicExponent} = PrivateKey = public_key:pem_entry_decode(PEMEntry).
#'RSAPrivateKey'{version = 'two-prime',
                 modulus = 22229493443645708705523581558090878957367586450909751107794238086962080015848054619397467933106360840581215574270745522166904017608027489665477935173052081890487447446310626920488171636893809207381146713861503360912041499077841558506105719930813981093031511085982222971766286284697738439495365308127623539168264485735318733143570874966874044788036410878275355326587265426041174941647279119175623719251367841389835668503095999900094770850870177317175179376777700441606775465664699045658323298203133653233096660719679323101930166930195443191172296135454691189674982978002129204335141598674600341635121716750790994857663,
                 publicExponent = 65537,
                 privateExponent = 9063156153229676924662253371869146981718142575465897883642248526536563742976639446881919269612615189287426646695978155123055302355715008680006262536032342464772946515181042026877091507664412194961994601437041210363149806298120549358120524841713071620699787543180874892131088843357547203920169690910021834483849134743555601203783807072109505533143770742728337175655038487885358069334754208953399452727458928811050452818025060994599655916130341740415252817366593987033651424125639494433027412278777046774479869728683039739604918658495070800181480018172418849752515334207484320393626124965634682208321416201446311633153,
                 prime1 = 164521523414867348932245957989117102915368563884993859002261566645808258344879910492420937163240199832331307100749119943409433286484548259511099355631330045213102436066823867425071151652762962497569166054934442815503467475819776236951618655403913958022234294442474216864427094473485506144195755649297015175787,
                 prime2 = 135116019972599473816048017685801033712236058170786861263736741860085553437437135622613712724374804076598166554483540794893509354227163240239139007233403699978088696404685225139789664570048411002777224365018069829297674225266705313820465805239182390476453868516395260463338929654977104656840607299316325284349,
                 exponent1 = 147569056142630304097428115330763687348780469954613104163891297033846417421386707636700983722231898117761929240130556277421611094690139690359651258394984594622991800078614740111927347586188229358333549863028003821758027253285773323663944810401203566052443974617489423747783424958868608468199693304366628623425,
                 exponent2 = 73698710132604687283553847265122443079654277300320703260674081988380277991486721807552564028842121569894176721899432882113283776882652613772807758847253949114496187350325506859820576764049629188340592937978560846360131842022199900104371441168882507779754893233678401144131966955711530283169525445605698764661,
                 coefficient = 54217906601613349900845012719441354443904142976126211011391339649360423968717271176279205528399064245082352536706243588002914706423805150006324479129342470987091660522921538081228784566290945826322084272584292895044988545674447894291015177179882935762102656930082543521724417149348501201209794855811489403546,
                 otherPrimeInfos = asn1_NOVALUE}
4> PublicKey = #'RSAPublicKey'{modulus=Modulus, publicExponent=PublicExponent}.
#'RSAPublicKey'{modulus = 22229493443645708705523581558090878957367586450909751107794238086962080015848054619397467933106360840581215574270745522166904017608027489665477935173052081890487447446310626920488171636893809207381146713861503360912041499077841558506105719930813981093031511085982222971766286284697738439495365308127623539168264485735318733143570874966874044788036410878275355326587265426041174941647279119175623719251367841389835668503095999900094770850870177317175179376777700441606775465664699045658323298203133653233096660719679323101930166930195443191172296135454691189674982978002129204335141598674600341635121716750790994857663,
                publicExponent = 65537}
5> Message = <<"hello world">>.
<<"hello world">>
6> Signature = crypto_rsassa_pss:sign(Message, sha256, PrivateKey).
<<38,90,46,243,100,214,118,52,91,101,152,27,12,35,237,53,
  226,149,135,146,212,87,237,40,107,225,154,69,195,...>>
7> true = crypto_rsassa_pss:verify(Message, sha256, Signature, PublicKey).
true
```


CHANGELOG.md
# Changelog

## 2.0.0 (2019-07-28)

* Enhancements
  * Support modern versions of OTP (20.0+).
  * Change license from MPL 2.0 to MIT.
  * Additional testing with PropEr and CAVP/FIPS test vectors.

## 1.0.0 (2015-07-20)

* Enhancements
  * Initial release.
  * Basic implementation based on https://github.com/vincenthz/hs-crypto-pubkey
  * Tested with ct and triq.


test/cavp_SUITE_data/hex.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
%%%-------------------------------------------------------------------
%%% @author Andrew Bennett <potatosaladx@gmail.com>
%%% @copyright 2014-2019, Andrew Bennett
%%% @doc
%%%
%%% @end
%%% Created :  12 Aug 2015 by Andrew Bennett <potatosaladx@gmail.com>
%%%-------------------------------------------------------------------
-module(hex).

%% API
-export([bin_to_hex/1]).
-export([hex_to_bin/1]).

-define(HEX_TO_INT(C),
	case C of
		$0 -> 16#0;
		$1 -> 16#1;
		$2 -> 16#2;
		$3 -> 16#3;
		$4 -> 16#4;
		$5 -> 16#5;
		$6 -> 16#6;
		$7 -> 16#7;
		$8 -> 16#8;
		$9 -> 16#9;
		$a -> 16#A;
		$b -> 16#B;
		$c -> 16#C;
		$d -> 16#D;
		$e -> 16#E;
		$f -> 16#F;
		$A -> 16#A;
		$B -> 16#B;
		$C -> 16#C;
		$D -> 16#D;
		$E -> 16#E;
		$F -> 16#F
	end).

-define(INT_TO_HEX(C),
	case C of
		16#0 -> $0;
		16#1 -> $1;
		16#2 -> $2;
		16#3 -> $3;
		16#4 -> $4;
		16#5 -> $5;
		16#6 -> $6;
		16#7 -> $7;
		16#8 -> $8;
		16#9 -> $9;
		16#A -> $a;
		16#B -> $b;
		16#C -> $c;
		16#D -> $d;
		16#E -> $e;
		16#F -> $f
	end).

%%====================================================================
%% API functions
%%====================================================================

bin_to_hex(Bin) ->
	<< << (?INT_TO_HEX(V bsr 4)), (?INT_TO_HEX(V band 16#F)) >> || << V >> <= Bin >>.

hex_to_bin(Hex) ->
	<< << ((?HEX_TO_INT(X) bsl 4) + ?HEX_TO_INT(Y)) >> || << X, Y >> <= Hex >>.

%%%-------------------------------------------------------------------
%%% Internal functions
%%%-------------------------------------------------------------------


test/crypto_rsassa_pss_ct.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
-module(crypto_rsassa_pss_ct).

-include_lib("common_test/include/ct.hrl").

%% API
-export([start/2]).
-export([stop/1]).
-export([progress_start/0]).
-export([progress_stop/1]).

%% Internal API
-export([init/1]).

%% Macros
-define(RED, "\e[0;31m").
-define(GREEN, "\e[0;32m").
-define(YELLOW, "\e[0;33m").
-define(WHITE, "\e[0;37m").
-define(CYAN, "\e[0;36m").
-define(RESET, "\e[0m").
-define(TIME, 1000). % timer:seconds(1)

start(Group, Config) ->
	Now = os:timestamp(),
	io:format(user, "~s[  ] ~s :: ~s~s", [?WHITE, format_utc_timestamp(Now), Group, ?RESET]),
	{ok, Progress} = progress_start(),
	[{crypto_rsassa_pss_ct, {Progress, Now}} | Config].

stop(Config) ->
	Now = os:timestamp(),
	{Progress, Old} = ?config(crypto_rsassa_pss_ct, Config),
	ok = progress_stop(Progress),
	Diff = timer:now_diff(Now, Old),
	io:format(user, "~s[OK] ~s :: ~s elapsed~s~n", [?GREEN, format_utc_timestamp(Now), format_elapsed_time(Diff), ?RESET]),
	ok.

progress_start() ->
	Ref = erlang:make_ref(),
	{ok, Pid} = proc_lib:start(?MODULE, init, [{self(), Ref}]),
	{ok, {Ref, Pid}}.

progress_stop({Ref, Pid}) ->
	Pid ! {stop, self(), Ref},
	receive
		Ref ->
			ok
	after
		1000 ->
			ok
	end.

%% @private
format_elapsed_time(USec) ->
	Micro = USec rem 1000000,
	Second = ((USec - Micro) div 1000000) rem 60,
	Minute = ((USec - (Second * 1000000)) div 6000000) rem 60,
	Hour = ((USec - (Minute * 6000000)) div 360000000) rem 24,
	io_lib:format("~2..0w:~2..0w:~2..0w.~6..0w", [Hour,Minute,Second,Micro]).

%% @private
format_utc_timestamp(TS = {_, _, Micro}) ->
	{{Year,Month,Day},{Hour,Minute,Second}} = calendar:now_to_universal_time(TS),
	io_lib:format("~4..0w-~2..0w-~2..0w ~2..0w:~2..0w:~2..0w.~6..0w", [Year,Month,Day,Hour,Minute,Second,Micro]).

%% @private
init({Parent, Ref}) ->
	process_flag(trap_exit, true),
	ok = proc_lib:init_ack(Parent, {ok, self()}),
	{ok, TRef} = timer:send_interval(?TIME, {tick, Ref}),
	loop(Ref, TRef).

%% @private
loop(Ref, TRef) ->
	receive
		{tick, Ref} ->
			io:format(user, "~s.~s", [?WHITE, ?RESET]),
			loop(Ref, TRef);
		{stop, Parent, Ref} when is_pid(Parent) ->
			io:format(user, "~n", []),
			catch timer:cancel(TRef),
			Parent ! Ref,
			exit(normal);
		Info ->
			io:format(user, "~n~s[~s] received unhandled message:~n~p~s~n", [?RED, ?MODULE, Info, ?RESET]),
			loop(Ref, TRef)
	end.


src/crypto_rsassa_pss.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
%%%-------------------------------------------------------------------
%%% @author Andrew Bennett <potatosaladx@gmail.com>
%%% @copyright 2014-2019, Andrew Bennett
%%% @doc PKCS #1: RSA Cryptography Specifications Version 2.1
%%% See RFC 3447: [https://tools.ietf.org/html/rfc3447]
%%% @end
%%% Created :  20 Jul 2015 by Andrew Bennett <potatosaladx@gmail.com>
%%%-------------------------------------------------------------------
-module(crypto_rsassa_pss).

-include_lib("public_key/include/public_key.hrl").

%% Public API
-export([sign/3]).
-export([sign/4]).
-export([verify/4]).
-export([verify/5]).
%% Private API
-export([emsa_pss_encode/3]).
-export([emsa_pss_encode/4]).
-export([emsa_pss_verify/4]).
-export([emsa_pss_verify/5]).
-export([mgf1/3]).
-export([rsassa_pss_sign/3]).
-export([rsassa_pss_sign/4]).
-export([rsassa_pss_verify/4]).
-export([rsassa_pss_verify/5]).

%% Types
-type rsa_digest_type() :: 'md5' | 'sha' | 'sha224' | 'sha256' | 'sha384' | 'sha512'.
-type rsa_hash_fun()    :: rsa_digest_type() | {hmac, rsa_digest_type(), iodata()} | fun((iodata()) -> binary()).
-type rsa_public_key()  :: #'RSAPublicKey'{}.
-type rsa_private_key() :: #'RSAPrivateKey'{}.

-define(PSS_TRAILER_FIELD, 16#BC).

%%====================================================================
%% Public API functions
%%====================================================================

-spec sign(Message, DigestType, RSAPrivateKey) -> Signature
	when
		Message       :: binary(),
		DigestType    :: rsa_digest_type(),
		RSAPrivateKey :: rsa_private_key(),
		Signature     :: binary().
sign(Message, DigestType, PrivateKey) ->
	sign(Message, DigestType, -2, PrivateKey).

-spec sign(Message, DigestType, Salt, RSAPrivateKey) -> Signature
	when
		Message       :: binary(),
		DigestType    :: rsa_digest_type(),
		Salt          :: integer() | binary(),
		RSAPrivateKey :: rsa_private_key(),
		Signature     :: binary().
sign(Message, DigestType, Salt, PrivateKey=#'RSAPrivateKey'{}) ->
	case rsassa_pss_sign(DigestType, Message, Salt, PrivateKey) of
		{ok, Signature} ->
			Signature;
		{error, Reason} ->
			erlang:error(Reason)
	end;
sign(Message, DigestType, Salt, PrivateKey) ->
	erlang:error(badarg, [Message, DigestType, Salt, PrivateKey]).

-spec verify(Message, DigestType, Signature, RSAPublicKey) -> boolean()
	when
		Message      :: binary(),
		DigestType   :: rsa_digest_type(),
		Signature    :: binary(),
		RSAPublicKey :: rsa_public_key().
verify(Message, DigestType, Signature, PublicKey) ->
	verify(Message, DigestType, Signature, -2, PublicKey).

-spec verify(Message, DigestType, Signature, Salt, RSAPublicKey) -> boolean()
	when
		Message      :: binary(),
		DigestType   :: rsa_digest_type(),
		Signature    :: binary(),
		Salt         :: integer() | binary(),
		RSAPublicKey :: rsa_public_key().
verify(Message, DigestType, Signature, Salt, PublicKey=#'RSAPublicKey'{}) ->
	rsassa_pss_verify(DigestType, Message, Signature, Salt, PublicKey);
verify(Message, DigestType, Signature, Salt, PublicKey) ->
	erlang:error(badarg, [Message, DigestType, Signature, Salt, PublicKey]).

%%====================================================================
%% Private API functions
%%====================================================================

%% See [https://tools.ietf.org/html/rfc3447#section-9.1.1]
-spec emsa_pss_encode(Hash, Message, EMBits) -> {ok, EM} | {error, Reason}
	when
		Hash    :: rsa_hash_fun(),
		Message :: binary(),
		EMBits  :: integer(),
		EM      :: binary(),
		Reason  :: term().
emsa_pss_encode(Hash, Message, EMBits)
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_integer(EMBits) ->
	emsa_pss_encode(Hash, Message, -2, EMBits);
emsa_pss_encode(Hash, Message, EMBits)
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	emsa_pss_encode(HashFun, Message, EMBits).

%% See [https://tools.ietf.org/html/rfc3447#section-9.1.1]
-spec emsa_pss_encode(Hash, Message, Salt, EMBits) -> {ok, EM} | {error, Reason}
	when
		Hash    :: rsa_hash_fun(),
		Message :: binary(),
		Salt    :: binary() | integer(),
		EMBits  :: integer(),
		EM      :: binary(),
		Reason  :: term().
emsa_pss_encode(Hash, Message, Salt, EMBits)
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_binary(Salt)
		andalso is_integer(EMBits) ->
	MHash = Hash(Message),
	HashLen = byte_size(MHash),
	SaltLen = byte_size(Salt),
	EMLen = ceiling(EMBits / 8),
	case EMLen < (HashLen + SaltLen + 2) of
		false ->
			MPrime = << 0:64, MHash/binary, Salt/binary >>,
			H = Hash(MPrime),
			PS = << 0:((EMLen - SaltLen - HashLen - 2) * 8) >>,
			DB = << PS/binary, 16#01, Salt/binary >>,
			case mgf1(Hash, H, EMLen - HashLen - 1) of
				{ok, DBMask} ->
					LeftBits = (EMLen * 8) - EMBits,
					<< _:LeftBits/bitstring, MaskedDBRight/bitstring >> = crypto:exor(DB, DBMask),
					MaskedDB = << 0:LeftBits, MaskedDBRight/bitstring >>,
					EM = << MaskedDB/binary, H/binary, ?PSS_TRAILER_FIELD >>,
					{ok, EM};
				MGF1Error ->
					MGF1Error
			end;
		true ->
			{error, encoding_error}
	end;
emsa_pss_encode(Hash, Message, -2, EMBits)
		when is_function(Hash, 1)
		andalso is_integer(EMBits) ->
	HashLen = byte_size(Hash(<<>>)),
	EMLen = ceiling(EMBits / 8),
	SaltLen = EMLen - HashLen - 2,
	case SaltLen < 0 of
		false ->
			emsa_pss_encode(Hash, Message, SaltLen, EMBits);
		true ->
			{error, encoding_error}
	end;
emsa_pss_encode(Hash, Message, -1, EMBits)
		when is_function(Hash, 1) ->
	HashLen = byte_size(Hash(<<>>)),
	SaltLen = HashLen,
	emsa_pss_encode(Hash, Message, SaltLen, EMBits);
emsa_pss_encode(Hash, Message, SaltLen, EMBits)
		when is_integer(SaltLen)
		andalso SaltLen >= 0 ->
	Salt = crypto:strong_rand_bytes(SaltLen),
	emsa_pss_encode(Hash, Message, Salt, EMBits);
emsa_pss_encode(Hash, Message, Salt, EMBits)
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	emsa_pss_encode(HashFun, Message, Salt, EMBits).

%% See [https://tools.ietf.org/html/rfc3447#section-9.1.2]
-spec emsa_pss_verify(Hash, Message, EM, EMBits) -> boolean()
	when
		Hash    :: rsa_hash_fun(),
		Message :: binary(),
		EM      :: binary(),
		EMBits  :: integer().
emsa_pss_verify(Hash, Message, EM, EMBits)
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_binary(EM)
		andalso is_integer(EMBits) ->
	emsa_pss_verify(Hash, Message, EM, -2, EMBits);
emsa_pss_verify(Hash, Message, EM, EMBits)
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	emsa_pss_verify(HashFun, Message, EM, EMBits).

%% See [https://tools.ietf.org/html/rfc3447#section-9.1.2]
-spec emsa_pss_verify(Hash, Message, EM, SaltLen, EMBits) -> boolean()
	when
		Hash    :: rsa_hash_fun(),
		Message :: binary(),
		EM      :: binary(),
		SaltLen :: integer(),
		EMBits  :: integer().
emsa_pss_verify(Hash, Message, EM, SaltLen, EMBits)
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_integer(SaltLen)
		andalso SaltLen >= 0
		andalso is_integer(EMBits) ->
	MHash = Hash(Message),
	HashLen = byte_size(MHash),
	EMLen = ceiling(EMBits / 8),
	MaskedDBLen = (EMLen - HashLen - 1),
	case {EMLen < (HashLen + SaltLen + 2), byte_size(EM), EM} of
		{false, EMLen, << MaskedDB:MaskedDBLen/binary, H:HashLen/binary, ?PSS_TRAILER_FIELD >>} ->
			LeftBits = ((EMLen * 8) - EMBits),
			case MaskedDB of
				<< 0:LeftBits, _/bitstring >> ->
					case mgf1(Hash, H, EMLen - HashLen - 1) of
						{ok, DBMask} ->
							<< _:LeftBits/bitstring, DBRight/bitstring >> = crypto:exor(MaskedDB, DBMask),
							DB = << 0:LeftBits, DBRight/bitstring >>,
							PSLen = ((EMLen - HashLen - SaltLen - 2) * 8),
							case DB of
								<< 0:PSLen, 16#01, Salt:SaltLen/binary >> ->
									MPrime = << 0:64, MHash/binary, Salt/binary >>,
									HPrime = Hash(MPrime),
									H =:= HPrime;
								_BadDB ->
									false
							end;
						_MGF1Error ->
							false
					end;
				_BadMaskedDB ->
					false
			end;
		_BadEMLen ->
			false
	end;
emsa_pss_verify(Hash, Message, EM, -2, EMBits)
		when is_function(Hash, 1)
		andalso is_integer(EMBits) ->
	HashLen = byte_size(Hash(<<>>)),
	EMLen = ceiling(EMBits / 8),
	SaltLen = EMLen - HashLen - 2,
	case SaltLen < 0 of
		false ->
			emsa_pss_verify(Hash, Message, EM, SaltLen, EMBits);
		true ->
			false
	end;
emsa_pss_verify(Hash, Message, EM, -1, EMBits)
		when is_function(Hash, 1) ->
	HashLen = byte_size(Hash(<<>>)),
	SaltLen = HashLen,
	emsa_pss_verify(Hash, Message, EM, SaltLen, EMBits).

%% See [https://tools.ietf.org/html/rfc3447#appendix-B.2]
-spec mgf1(Hash, Seed, MaskLen) -> {ok, binary()} | {error, mask_too_long}
	when
		Hash    :: rsa_hash_fun(),
		Seed    :: binary(),
		MaskLen :: pos_integer().
mgf1(Hash, Seed, MaskLen)
		when is_function(Hash, 1)
		andalso is_binary(Seed)
		andalso is_integer(MaskLen)
		andalso MaskLen >= 0 ->
	HashLen = byte_size(Hash(<<>>)),
	case MaskLen > (16#FFFFFFFF * HashLen) of
		false ->
			Reps = ceiling(MaskLen / HashLen),
			{ok, derive_mgf1(Hash, 0, Reps, Seed, MaskLen, <<>>)};
		true ->
			{error, mask_too_long}
	end;
mgf1(Hash, Seed, MaskLen)
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	mgf1(HashFun, Seed, MaskLen).

%% See [https://tools.ietf.org/html/rfc3447#section-8.1.1]
-spec rsassa_pss_sign(Hash, Message, RSAPrivateKey) -> {ok, Signature} | {error, Reason}
	when
		Hash          :: rsa_hash_fun(),
		Message       :: binary(),
		RSAPrivateKey :: rsa_private_key(),
		Signature     :: binary(),
		Reason        :: term().
rsassa_pss_sign(Hash, Message, RSAPrivateKey=#'RSAPrivateKey'{modulus=Modulus})
		when is_function(Hash, 1)
		andalso is_binary(Message) ->
	ModBits = int_to_bit_size(Modulus),
	case emsa_pss_encode(Hash, Message, ModBits - 1) of
		{ok, EM} ->
			ModBytes = int_to_byte_size(Modulus),
			S = pad_to_key_size(ModBytes, dp(EM, RSAPrivateKey)),
			{ok, S};
		EncodingError ->
			EncodingError
	end;
rsassa_pss_sign(Hash, Message, RSAPrivateKey=#'RSAPrivateKey'{})
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	rsassa_pss_sign(HashFun, Message, RSAPrivateKey).

%% See [https://tools.ietf.org/html/rfc3447#section-8.1.1]
-spec rsassa_pss_sign(Hash, Message, Salt, RSAPrivateKey) -> {ok, Signature} | {error, Reason}
	when
		Hash          :: rsa_hash_fun(),
		Message       :: binary(),
		Salt          :: binary() | integer(),
		RSAPrivateKey :: rsa_private_key(),
		Signature     :: binary(),
		Reason        :: term().
rsassa_pss_sign(Hash, Message, Salt, RSAPrivateKey=#'RSAPrivateKey'{modulus=Modulus})
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso (is_binary(Salt) orelse is_integer(Salt)) ->
	ModBits = int_to_bit_size(Modulus),
	case emsa_pss_encode(Hash, Message, Salt, ModBits - 1) of
		{ok, EM} ->
			ModBytes = int_to_byte_size(Modulus),
			S = pad_to_key_size(ModBytes, dp(EM, RSAPrivateKey)),
			{ok, S};
		EncodingError ->
			EncodingError
	end;
rsassa_pss_sign(Hash, Message, Salt, RSAPrivateKey=#'RSAPrivateKey'{})
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	rsassa_pss_sign(HashFun, Message, Salt, RSAPrivateKey).

%% See [https://tools.ietf.org/html/rfc3447#section-8.1.2]
-spec rsassa_pss_verify(Hash, Message, Signature, RSAPublicKey) -> boolean()
	when
		Hash         :: rsa_hash_fun(),
		Message      :: binary(),
		Signature    :: binary(),
		RSAPublicKey :: rsa_public_key().
rsassa_pss_verify(Hash, Message, Signature, RSAPublicKey=#'RSAPublicKey'{modulus=Modulus})
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_binary(Signature) ->
	ModBytes = int_to_byte_size(Modulus),
	case byte_size(Signature) =:= ModBytes of
		true ->
			ModBits = int_to_bit_size(Modulus),
			EM = pad_to_key_size(ceiling((ModBits - 1) / 8), ep(Signature, RSAPublicKey)),
			emsa_pss_verify(Hash, Message, EM, ModBits - 1);
		false ->
			false
	end;
rsassa_pss_verify(Hash, Message, Signature, RSAPublicKey=#'RSAPublicKey'{})
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	rsassa_pss_verify(HashFun, Message, Signature, RSAPublicKey).

%% See [https://tools.ietf.org/html/rfc3447#section-8.1.2]
-spec rsassa_pss_verify(Hash, Message, Signature, SaltLen, RSAPublicKey) -> boolean()
	when
		Hash         :: rsa_hash_fun(),
		Message      :: binary(),
		Signature    :: binary(),
		SaltLen      :: integer(),
		RSAPublicKey :: rsa_public_key().
rsassa_pss_verify(Hash, Message, Signature, SaltLen, RSAPublicKey=#'RSAPublicKey'{modulus=Modulus})
		when is_function(Hash, 1)
		andalso is_binary(Message)
		andalso is_binary(Signature)
		andalso is_integer(SaltLen) ->
	ModBytes = int_to_byte_size(Modulus),
	case byte_size(Signature) =:= ModBytes of
		true ->
			ModBits = int_to_bit_size(Modulus),
			EM = pad_to_key_size(ceiling((ModBits - 1) / 8), ep(Signature, RSAPublicKey)),
			emsa_pss_verify(Hash, Message, EM, SaltLen, ModBits - 1);
		false ->
			false
	end;
rsassa_pss_verify(Hash, Message, Signature, SaltLen, RSAPublicKey=#'RSAPublicKey'{})
		when is_tuple(Hash)
		orelse is_atom(Hash) ->
	HashFun = resolve_hash(Hash),
	rsassa_pss_verify(HashFun, Message, Signature, SaltLen, RSAPublicKey).

%%%-------------------------------------------------------------------
%%% Internal functions
%%%-------------------------------------------------------------------

%% @private
ceiling(X) when X < 0 ->
	trunc(X);
ceiling(X) ->
	T = trunc(X),
	case X - T == 0 of
		false ->
			T + 1;
		true ->
			T
	end.

%% @private
derive_mgf1(_Hash, Reps, Reps, _Seed, MaskLen, T) ->
	binary:part(T, 0, MaskLen);
derive_mgf1(Hash, Counter, Reps, Seed, MaskLen, T) ->
	CounterBin = << Counter:8/unsigned-big-integer-unit:4 >>,
	NewT = << T/binary, (Hash(<< Seed/binary, CounterBin/binary >>))/binary >>,
	derive_mgf1(Hash, Counter + 1, Reps, Seed, MaskLen, NewT).

%% @private
dp(B, #'RSAPrivateKey'{modulus=N, privateExponent=E}) ->
	crypto:mod_pow(B, E, N).

%% @private
ep(B, #'RSAPublicKey'{modulus=N, publicExponent=E}) ->
	crypto:mod_pow(B, E, N).

%% @private
int_to_bit_size(I) ->
	int_to_bit_size(I, 0).

%% @private
int_to_bit_size(0, B) ->
	B;
int_to_bit_size(I, B) ->
	int_to_bit_size(I bsr 1, B + 1).

%% @private
int_to_byte_size(I) ->
	int_to_byte_size(I, 0).

%% @private
int_to_byte_size(0, B) ->
	B;
int_to_byte_size(I, B) ->
	int_to_byte_size(I bsr 8, B + 1).

%% @private
pad_to_key_size(Bytes, Data) when byte_size(Data) < Bytes ->
	pad_to_key_size(Bytes, << 0, Data/binary >>);
pad_to_key_size(_Bytes, Data) ->
	Data.

%% @private
resolve_hash(HashFun) when is_function(HashFun, 1) ->
	HashFun;
resolve_hash(DigestType) when is_atom(DigestType) ->
	fun(Data) ->
		crypto:hash(DigestType, Data)
	end;
resolve_hash({hmac, DigestType, Key}) when is_atom(DigestType) ->
	fun(Data) ->
		crypto:hmac(DigestType, Key, Data)
	end.
